### ISSUE TITLE
Check init executable for quality

### ISSUE THREAD

**Issue by sanmai-NL** [hadolint/hadolint#909]
- [ ] This is a bug report
- [x] This is a feature request
- [x] I searched existing issues before opening this one

Many container images a built without a proper init process as PID 1. [This issue has come up in the Kubernetes community](https://github.com/kubernetes/kubernetes/issues/84210).

Rather than or beyond changing Kubernetes to, similarly to Docker Engine, inject an init executable, I propose that Hadolint evaluates the resulting entrypoint for whether it is a ‘known good’ init executable, just based on its file path. 
I don't think a more sophisticated check adds value, since there is little diversity in popular init executables in practice. If some special init executable is used, the warning can be silenced anyway.

**Comment by m-ildefons** [comment]
Hi Sander,

This is an interesting issue, but I'm not sure this would translate very well into a rule. There are many cases where this rule would not apply, almost all of which can't be detected by looking just at the Dockerfile, which is what Hadolint is doing.
Depending on the use case it can also be unnecessary to build the init process into the container, as e.g. Docker can inject one at runtime.
I also think that the solution with the init process does more harm than good in most cases, when processes don't spawn off other processes, when they run to completion on their own quickly (and thus usually doesn't need signal handling to be set up) or when the binaries executed as PID 1 already implement the correct behaviour. Checking for these is out of scope for Hadolint, as the information needed can't be determined from the Dockerfile alone.
Not to mention that the recommendations from Docker aren't so clear in favor of using an init process at all times:
https://github.com/docker-library/official-images#init

It would be a fun idea to have sort of a 'test-suite' for K8s readiness for containers though, where containers are checked for properties like correct behaviour on system signals or a zombie process is injected into a container to see if whatever runs as PID 1 cleans it up.

**Comment by sanmai-NL** [comment]
> Hi Sander,
> 
> 
> 
> This is an interesting issue, but I'm not sure this would translate very well into a rule. There are many cases where this rule would not apply, almost all of which can't be detected by looking just at the Dockerfile, which is what Hadolint is doing.

I'm curious. Can you give representative exampkes?

> Depending on the use case it can also be unnecessary to build the init process into the container, as e.g. Docker can inject one at runtime.

Depending on the use case, e.g., specifying limiting software version ranges can be inapplicable as well. I think the benchmark should be: does a recommendation add more value than harm? A 'redundant' init executable doesn't harm.

> I also think that the solution with the init process does more harm than good in most cases, when processes don't spawn off other processes, when they run to completion on their own quickly (and thus usually doesn't need signal handling to be set up) or when the binaries executed as PID 1 already implement the correct behaviour. Checking for these is out of scope for Hadolint, as the information needed can't be determined from the Dockerfile alone.

In what sense would a tiny init executable do harm here?


> Not to mention that the recommendations from Docker aren't so clear in favor of using an init process at all times:
> 
> https://github.com/docker-library/official-images#init


Okay, I'm reading this very differently. This document gives two compelling reasons to use a proper init process. It doesn't suggest when these reasons don't apply (which I think is _never_, by the way) that a proper init process is inadvisable.

> It would be a fun idea to have sort of a 'test-suite' for K8s readiness for containers though, where containers are checked for properties like correct behaviour on system signals or a zombie process is injected into a container to see if whatever runs as PID 1 cleans it up.

Similar to shellcheck expecting a specification of which shell executable is being used, why not require or suggest this with Hadolint?

When Dockerfile is named Containerfile, the Docker feature you refer to is out of the equation statically.

**Comment by m-ildefons** [comment]
Hi Sander,

one example for when I think it would be hard to make a good recommendation is:

```Dockerfile
FROM debian:buster
COPY --chmod=755 entrypoint.sh /entrypoint.sh
CMD [ "/entrypoint.sh" ]
```

Technically the entrypoint is not an init-like process, but we can't tell what the entrypoint script is doing. Maybe it `exec`s into a proper init process. Maybe the base image already specifies something like `tini` as `ENTRYPOINT`. Maybe the script is executed with an interpreter that covers the signal handling and process cleanup already. Just from looking at the source of the Dockerfile this will always result situations with false positives.
And not to forget that applications can implement the signal handling and cleanup themselves too, so they may not need an init process anyways, which is especially likely with applications engineered specifically to run inside of containers.

A "tiny" init executable is IMO doing harm in the sense that it increases the image size, installing also may create an additional layer and it does consume resources at runtime. Not much but a stack still needs to be maintained and a couple hundred instructions still need to be executed before the real application launches.
It also violates the recommendation of one process per container:

> predominantly this means just one process per container, so there is no need for a full init system.

from the official docker guidelines. Lastly it adds another failure point in the mix.

So I still don't see how a general recommendation can be made based on the Dockerfile alone.
